src/Common/HTTPConnectionPool.cpp

```diff
@@ -393,6 +453,8 @@ class EndpointConnectionPool : public std::enable_shared_from_this<EndpointConne
                 }
             }
             response_stream = nullptr;
+            Session::setSendDataHooks();
+            Session::setReceiveDataHooks();
```

Comment:
This reset lead to a data race: https://pastila.nl/?000777c6/1c25eebbf78952d5ed8d7c42cbcca009#or9ialwCHfjZi17MNTG5Hg==

So we cannot reset it here, because session could be still in use for some reason. I think I should return to the logic I used a few revisions earlier: reset hooks when the connection is reassigned to another request in `sendRequest()`.

Reply:
It turned out to be correct. Data race was inside ResourceGuard